src/cron/arweave.rs

```diff
@@ -159,25 +184,53 @@ impl Arweave {
         &self,
         ctx: &Context,
-    ) -> reqwest::Result<NetworkInfo>
+    ) -> Result<NetworkInfo, HttpClient::Error>
     where
         Context: ArweaveContext<HttpClient>,
         HttpClient: Client<Request = reqwest::Request, Response = reqwest::Response>,
+        HttpClient::Error: From<reqwest::Error>,
     {
         info!("Fetch network info");
         let uri = http::uri::Uri::from_str(&format!("{}info", self.get_host())).unwrap();
-        let req: http::Request<String> = http::request::Builder::new()
-            .method(http::Method::GET)
-            .uri(uri)
-            .body("".to_string())
-            .unwrap();
 
-        let req: reqwest::Request = reqwest::Request::try_from(req).unwrap();
-        let res: reqwest::Response = ctx.get_client().execute(req).await.expect("request failed"); // FIXME: should not panic, handle failure
-        if res.status().is_success() {
-            return res.json().await;
-        } else {
-            Err(res.error_for_status().err().unwrap()) // FIXME: do not unwrap
+        let client = ctx.get_client();
+        let res = retry::<tokio::runtime::Handle, _>()
+            .max_retries(3)
+            .run_with_context(&(client, uri), |(client, uri)| async move {
+                let req: http::Request<String> = http::request::Builder::new()
+                    .method(http::Method::GET)
+                    .uri(uri)
+                    .body("".to_string())
+                    .unwrap();
+                let req: reqwest::Request = reqwest::Request::try_from(req).unwrap();
+                let res = client.execute(req).await;
+                match res {
+                    Ok(res) => {
+                        if let Some(retry_after) = res.headers().get(http::header::RETRY_AFTER) {
+                            let retry_delay = match retry_after.to_str().unwrap().parse().unwrap() {
+                                RetryAfter::Duration(duration_seconds) => {
+                                    Duration::seconds(duration_seconds).to_std().unwrap()
+                                }
+                                RetryAfter::Timestamp(timestamp) => timestamp
+                                    .signed_duration_since(Utc::now())
+                                    .to_std()
+                                    .unwrap(),
+                            };
+                            RetryControl::Retry(Ok(res), Some(retry_delay))
+                        } else if res.status().is_server_error() {
+                            RetryControl::Retry(Ok(res), None)
+                        } else {
+                            RetryControl::Success(Ok(res))
+                        }
+                    }
+                    ret @ Err(_) => RetryControl::Fail(ret),
+                }
+            })
+            .await?;
+
+        match res.error_for_status() {
+            Ok(res) => res.json().await.map_err(|err| err.into()),
+            Err(err) => Err(err.into()),
```

Comment:
This is the main part of the change, repeating the request when receiving `Retry-After` header or server error.

Reply:
This could be encapsulated into a function to make the code a bit more readable and to allow reusing this retry logic. I'll do that if we agree over the logic and also if we need the retry in some other parts of the code.

Reply:
Looks good. It would be nice to have this retry logic, with a `max_retries` parameter, separated and reusable, for sure.